src/components/AttachmentPicker/index.native.js

```diff
@@ -221,19 +229,16 @@ class AttachmentPicker extends Component {
                     onClose={this.close}
                     isVisible={this.state.isVisible}
                     anchorPosition={styles.createMenuPosition}
-                    onModalHide={this.completeAttachmentSelection}
+                    onModalHide={this.onModalHide}
```

Comment:
I'm not sure I follow what is going on here.

Correct me if I'm wrong on any of the following:

1. `this.onModalHide` is defined when `this.selectItem()` is called
2. `this.onModalHide` is passed as the `onModalHide` prop to a `Popover` element.
3. `this.selectItem()` gets called (which in turn defines `this.onModalHide`) after we pass `this.onModalHide` as a prop to the `Popover` element.

I fear we might be in some sort of race condition here...

Reply:
It's very similar to the handling in `CreateMenu`: https://github.com/Expensify/Expensify.cash/blob/main/src/components/CreateMenu/index.native.js#L15-L24

1. Something external will trigger this attachment picker modal to show (Pressing "Add Attachment" on the previous modal)
2. This modal - `AttachmentPicker` appears and lists the 3 options: camera/galler/documents
2. `selectItem` is triggered when the users taps on an option
3. When this happens we're done with the modal  hence the `close` call
4. We want to launch the camera/gallery only when the modal is fully closed and so we capture the `onModalHide` callback
5. The modal closes and the captured callback is executed
6. camera/gallery is launched
7. No matter if the user selects or cancels on `finally` the captured callback is cleared

Even if the modal closes immediately `onModalHide` is called correctly
After `selectItem` is called it's impossible to call it again until the modal is opened again and you start over from 1)

Basically `setItem` captures a callback to be called after close and then calls `close`. The call to `close` triggers a re-render and the `onModalHide` is applied to the popover: both the `isVisible` and `onModalHide` are always applied at the same time on the `render` that is triggered by `setState` 

Sorry if I've over explained this